GpuMiner/MinerEngine/CLMiner.cpp

```diff
@@ -462,19 +473,33 @@ bool CLMiner::Initialize()
 
         // create miner OpenCL program
         cl::Program::Sources sources { { _kernelCode.data(), _kernelCode.size() } };
-        cl::Program program(_context, sources);
+        cl::Program program(_context, sources);
         try
-        {
+        {
             program.build({ device }, options);
             cllog << "Build info:" << program.getBuildInfo<CL_PROGRAM_BUILD_LOG>(device);
         }
         catch(cl::Error const&)
         {
             cwarn << "Build info:" << program.getBuildInfo<CL_PROGRAM_BUILD_LOG>(device);
             return false;
-        }
-
-        _searchKernel = cl::Kernel(program, "search_nonce");
+        }
+
+
+        _searchKernel = cl::Kernel(program, "search_nonce");
+        
+#if defined (__APPLE__) || defined (__MACOS)
+        size_t local;
+        
+        int err;
+        err = clGetKernelWorkGroupInfo(_searchKernel.get(), device.get(), CL_KERNEL_WORK_GROUP_SIZE, sizeof(local), &local, NULL);
+        if (err != CL_SUCCESS) {
+            fprintf(stdout, "Error: Failed to retrieve kernel work group info! err: %d\n", err);
+            return false;
+        }
+        
+        _workgroupSize = (uint32_t)local;
```

Comment:
Why do you need it?

Reply:
The workgroup size should use the opencl device workgroup, and on mac it has problem since the workgroupSize was modified to align with 8. I think we should compare the default for device, and use the min one, but i need verify this thought on Linux and Windows later, then we can improve this.

Reply:
I suggest to place it before line 444
if(_globalWorkSize % _workgroupSize != 0)
        {
            _globalWorkSize = ((_globalWorkSize / _workgroupSize) + 1) * _workgroupSize;
        }

// make sure that global work size is evenly divisible by the local workgroup size

Reply:
Good point. Maybe move this part to line 501, since the kernel is just compiled, and the local workgroup size if fetched after kernel compiled. What do you think?